tokio/src/io/lockless_split.rs

```diff
@@ -0,0 +1,178 @@
+use std::{
+    cell::UnsafeCell,
+    error::Error,
+    fmt::{self, Debug},
+    io,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use super::{AsyncRead, AsyncWrite, ReadBuf};
+
+cfg_net_or_io_util! {
+    /// Owned Read Half Part
+    #[derive(Debug)]
+    pub struct LocklessOwnedReadHalf<T>(pub Arc<UnsafeCell<T>>);
+    /// Owned Write Half Part
+    #[derive(Debug)]
+    pub struct LocklessOwnedWriteHalf<T>(pub Arc<UnsafeCell<T>>)
+    where
+        T: Shutdown;
+
+    unsafe impl<T: Send + Sync> Sync for LocklessOwnedReadHalf<T> {}
+    unsafe impl<T: Send + Sync> Send for LocklessOwnedReadHalf<T> {}
+    unsafe impl<T: Send + Sync + Shutdown> Sync for LocklessOwnedWriteHalf<T> {}
+    unsafe impl<T: Send + Sync + Shutdown> Send for LocklessOwnedWriteHalf<T> {}
+
+    /// Inner split trait
+    pub trait LocklessSplitable {
+        /// Owned Read Split
+        type OwnedReadHalf;
+        /// Owned Write Split
+        type OwnedWriteHalf;
+
+        /// Split into owned parts
+        fn into_split(self) -> (Self::OwnedReadHalf, Self::OwnedWriteHalf);
+    }
+
+    /// The object with has this `LocklessSplit` trait can be safely split
+    /// to read/write object in both form of `Owned` or `Borrowed`.
+    ///
+    /// # Safety
+    ///
+    /// Users should ensure the read
+    /// operations are indenpendent from the write ones, the methods
+    /// from `AsyncRead` and `AsyncWrite` can execute concurrently.
+    pub unsafe trait LocklessSplit {}
```

Comment:
This is basically saying "create multiple mutable references to me!", which is wildly unsafe and not going to happen.